File: python/src/bound_entangled/cd_otimes_cd/orthogonal_singlet.py
"""https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.023101 as PPT singlets.
ρ_F2 from Phys. Rev. Research 3, 023101 (2021) — the "second family" of
PPT singlets, reverse-engineered from Tóth & Vértesi, PRL 120, 020506 (2018).
Port of BES_metro.m from the QUBIT4MATLAB package."""

from math import sqrt, log2
import numpy as np
from toqito.matrix_ops import tensor
from bound_entangled.utils import ketbra


def ket_i(d: int, i: int):
    ket = np.zeros(d, dtype=np.complex128)
    ket[i] = 1
    return ket


def basis(shield_dim: int, i: int, j: int, k: int, l: int):
    """basis vector |i j k l> in C^2 ⊗ C^2 ⊗ C^d ⊗ C^d"""
    assert i <= 1 and j <= 1
    return tensor(ket_i(2, i), ket_i(2, j), ket_i(shield_dim, k), ket_i(shield_dim, l))


def _build_Q_matrices(d: int) -> np.ndarray:
    """Orthogonal matrices Q^k from Appendix G. Returns shape (d, d, d)
    with Q[k] = Q^k.  d must be 3 or a power of 2."""
    if d == 3:
        Q = np.zeros((3, 3, 3), dtype=np.complex128)
        for k in range(3):
            phi = 2 * np.pi * (k + 1) / 3
            c, s = np.cos(phi), np.sin(phi)
            Q[k] = np.array([[c, s, 0], [s, -c, 0], [0, 0, 1]], dtype=np.complex128)
        return Q

    if d >= 2 and (d & (d - 1)) == 0:
        n = int(round(log2(d)))
        X = np.array([[0, 1], [1, 0]], dtype=np.complex128)
        I2 = np.eye(2, dtype=np.complex128)
        Q = np.zeros((d, d, d), dtype=np.complex128)
        for k in range(d):
            bits = format(k, f"0{n}b")
            factors = [X if b == "1" else I2 for b in bits]
            Pk = factors[0]
            for f in factors[1:]:
                Pk = np.kron(Pk, f)
            Q[k] = Pk
        return Q

    raise ValueError("d must be 3 or a power of 2 (so D = 2d ∈ {4, 6, 8, 16, ...}).")


def _build_sk_vectors(Q: np.ndarray, d: int):
    """The |s_k> vectors on A'B' that pair with |01>_AB in sum2.
    For d = 3 the paper hard-codes them (Eqs. G4-G5, symmetric partners of |t_k>);
    for d = 2^n we take |s_k> = |t_k> per Eq. G16."""
    if d == 3:
        e = [ket_i(d, i) for i in range(d)]
        s0 = (tensor(e[0], e[0]) + tensor(e[1], e[1])) / sqrt(2)
        s1 = (tensor(e[0], e[1]) - tensor(e[1], e[0])) / sqrt(2)
        s2 = tensor(e[2], e[2])
        return [s0, s1, s2]

    sk = []
    for k in range(d):
        t = np.zeros(d * d, dtype=np.complex128)
        for i in range(d):
            for j in range(d):
                if Q[k, i, j] != 0:
                    t = t + Q[k, i, j] * tensor(ket_i(d, i), ket_i(d, j))
        sk.append(t / sqrt(d))
    return sk


def orthogonal_singlet(*, shield_dim: int) -> np.ndarray:
    """Construct the ρ_F2 (second-family) PPT singlet on C^{2d} ⊗ C^{2d}.

    Reverse-engineered from Tóth & Vértesi, PRL 120, 020506 (2018) and
    presented as ρ_F2 in Phys. Rev. Research 3, 023101 (2021).  The state is
    bound entangled for every valid shield_dim.  The returned matrix is in
    ABA'B' ordering (two-qubit pair AB, shield pair A'B').

    Args:
        shield_dim: shield subsystem dimension d.  Must be 3 or a power of 2
            (equivalently, total local dimension 2d ∈ {4, 6, 8, 16, …}).

    Returns:
        np.ndarray: (4d²) × (4d²) density matrix of the ρ_F2 bound-entangled state.

    Raises:
        ValueError: If shield_dim is not 3 or a power of 2.
    """
    d = shield_dim
    p1 = sqrt(d) / (1 + sqrt(d))
    p2 = 1 - p1

    Q = _build_Q_matrices(d)
    sk_vecs = _build_sk_vectors(Q, d)

    sum1 = np.zeros((4 * d * d, 4 * d * d), dtype=np.complex128)
    for i in range(d):
        for j in range(d):
            z = basis(d, 0, 0, i, j) / sqrt(2)
            for k in range(d):
                if Q[j, i, k] != 0:
                    z = z + (Q[j, i, k] / sqrt(2)) * basis(d, 1, 1, j, k)
            sum1 = sum1 + ketbra(z)

    ab01 = np.zeros(4, dtype=np.complex128)
    ab01[1] = 1.0
    sum2 = np.zeros((4 * d * d, 4 * d * d), dtype=np.complex128)
    for k in range(d):
        sum2 = sum2 + np.kron(ketbra(ab01), ketbra(sk_vecs[k]))

    sum3 = np.zeros((4 * d * d, 4 * d * d), dtype=np.complex128)
    for i in range(d):
        v = basis(d, 1, 0, i, i)
        sum3 = sum3 + ketbra(v)

    return (p1 / d**2) * sum1 + (p2 / (2 * d)) * sum2 + (p2 / (2 * d)) * sum3

File: js/src/cdOtimesCd/orthogonalSinglet.ts
/**
 * https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.023101
 *
 * ρ_F2 from Phys. Rev. Research 3, 023101 (2021) — the "second family" of PPT
 * singlets, reverse-engineered from G. Tóth, T. Vértesi, "Quantum States with a
 * Positive Partial Transpose are Useful for Metrology", Phys. Rev. Lett. 120,
 * 020506 (2018). Port of `BES_metro.m` from the QUBIT4MATLAB package.
 */

import { multiply } from 'mathjs';
import type { Matrix } from 'mathjs';
import { ket, ketbra, permuteSystems, tensor } from '../utils/index.js';
import { fail, sumMatrices } from '../utils/internal.js';

export interface OrthogonalSingletOptions {
  /**
   * The shield subsystem dimension d. Must be 3 or a power of two, so that the
   * total local dimension 2d is one of 4, 6, 8, 16, ...
   */
  shieldDim: number;
}

/** The product basis state |i j k l> of C² ⊗ C² ⊗ C^d ⊗ C^d. */
function basis(d: number, i: number, j: number, k: number, l: number): Matrix {
  return ket([2, 2, d, d], [i, j, k, l]);
}

/**
 * The orthogonal matrices Q^k of Appendix G, as `Q[k][row][column]`.
 *
 * For d = 3 these are the reflections/rotations of Eq. G3; for d = 2^n they are
 * the 2^n tensor products of the Pauli X and the identity, one per bit pattern.
 */
function buildQMatrices(d: number): number[][][] {
  if (d === 3) {
    return [0, 1, 2].map((k) => {
      const phi = (2 * Math.PI * (k + 1)) / 3;
      const c = Math.cos(phi);
      const s = Math.sin(phi);
      return [
        [c, s, 0],
        [s, -c, 0],
        [0, 0, 1],
      ];
    });
  }

  if (d >= 2 && (d & (d - 1)) === 0) {
    const n = Math.round(Math.log2(d));
    const x = [
      [0, 1],
      [1, 0],
    ];
    const identity = [
      [1, 0],
      [0, 1],
    ];
    const matrices: number[][][] = [];
    for (let k = 0; k < d; k++) {
      const bits = k.toString(2).padStart(n, '0');
      let product = bits[0] === '1' ? x : identity;
      for (const bit of bits.slice(1)) {
        const factor = bit === '1' ? x : identity;
        const size = product.length;
        const next: number[][] = [];
        for (let row = 0; row < size * 2; row++) {
          next.push(new Array<number>(size * 2).fill(0));
          for (let column = 0; column < size * 2; column++) {
            next[row][column] =
              product[Math.floor(row / 2)][Math.floor(column / 2)] * factor[row % 2][column % 2];
          }
        }
        product = next;
      }
      matrices.push(product);
    }
    return matrices;
  }

  return fail(`the shield dimension must be 3 or a power of two, got ${d}`);
}

/**
 * The |s_k> vectors on A'B' that pair with |01>_AB.
 *
 * For d = 3 the paper hard-codes them (Eqs. G4-G5, the symmetric partners of
 * the |t_k>); for d = 2^n they are the |t_k> themselves (Eq. G16).
 */
function buildSkVectors(q: number[][][], d: number): Matrix[] {
  if (d === 3) {
    const e = (i: number): Matrix => ket(d, i);
    return [
      multiply(1 / Math.SQRT2, sumMatrices([tensor(e(0), e(0)), tensor(e(1), e(1))])) as Matrix,
      multiply(
        1 / Math.SQRT2,
        sumMatrices([tensor(e(0), e(1)), multiply(-1, tensor(e(1), e(0))) as Matrix]),
      ) as Matrix,
      tensor(e(2), e(2)),
    ];
  }

  const vectors: Matrix[] = [];
  for (let k = 0; k < d; k++) {
    const terms: Matrix[] = [];
    for (let i = 0; i < d; i++) {
      for (let j = 0; j < d; j++) {
        if (q[k][i][j] !== 0) {
          terms.push(multiply(q[k][i][j], ket([d, d], [i, j])) as Matrix);
        }
      }
    }
    vectors.push(multiply(1 / Math.sqrt(d), sumMatrices(terms)) as Matrix);
  }
  return vectors;
}

/**
 * The ρ_F2 (second-family) PPT singlet on C^2d ⊗ C^2d.
 *
 * Reverse-engineered from Tóth & Vértesi and presented as ρ_F2 in Phys. Rev.
 * Research 3, 023101 (2021). Bound entangled for every valid `shieldDim`. The
 * construction is built up in ABA'B' ordering (two-qubit pair AB, shield pair
 * A'B'), then permuted to AA'BB' before being returned, so the
 * Alice = (A, A') | Bob = (B, B') cut can be tested directly on the result.
 *
 * @param options - see {@link OrthogonalSingletOptions}.
 * @returns the (4d²) x (4d²) density matrix of the ρ_F2 singlet, in AA'BB'
 * ordering.
 */
export function orthogonalSinglet({ shieldDim }: OrthogonalSingletOptions): Matrix {
  const d = shieldDim;
  if (!Number.isInteger(d) || d < 2) {
    fail(`the shield dimension must be an integer >= 2, got ${d}`);
  }

  const p1 = Math.sqrt(d) / (1 + Math.sqrt(d));
  const p2 = 1 - p1;

  const q = buildQMatrices(d);
  const skVectors = buildSkVectors(q, d);

  // sum1: the |00 ij> component paired, through Q^j, with the |11 jk> ones.
  const sum1Terms: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    for (let j = 0; j < d; j++) {
      const terms: Matrix[] = [multiply(1 / Math.SQRT2, basis(d, 0, 0, i, j)) as Matrix];
      for (let k = 0; k < d; k++) {
        if (q[j][i][k] !== 0) {
          terms.push(multiply(q[j][i][k] / Math.SQRT2, basis(d, 1, 1, j, k)) as Matrix);
        }
      }
      sum1Terms.push(ketbra(sumMatrices(terms)));
    }
  }

  // sum2: |01>_AB tensored with each of the |s_k>.
  const ab01 = ket([2, 2], [0, 1]);
  const sum2Terms = skVectors.map((sk) => ketbra(tensor(ab01, sk)));

  // sum3: the |10 ii> populations.
  const sum3Terms: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    sum3Terms.push(ketbra(basis(d, 1, 0, i, i)));
  }

  const rho = sumMatrices([
    multiply(p1 / d ** 2, sumMatrices(sum1Terms)) as Matrix,
    multiply(p2 / (2 * d), sumMatrices(sum2Terms)) as Matrix,
    multiply(p2 / (2 * d), sumMatrices(sum3Terms)) as Matrix,
  ]);
  return permuteSystems(rho, [0, 2, 1, 3], [2, 2, d, d]);
}
